server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java

```diff
@@ -125,38 +125,42 @@ public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAl
 
     @Override
     public Decision shouldAutoExpandToNode(IndexMetadata indexMetadata, DiscoveryNode node, RoutingAllocation allocation) {
-        Decision decision = shouldClusterFilter(node, allocation);
+        Decision decision = shouldIndexFilter(indexMetadata, node, allocation);
         if (decision != null) return decision;
 
-        decision = shouldIndexFilter(indexMetadata, node, allocation);
+        decision = shouldClusterFilter(node, allocation);
         if (decision != null) return decision;
 
         return allocation.decision(Decision.YES, NAME, "node passes include/exclude/require filters");
     }
 
     private Decision shouldFilter(ShardRouting shardRouting, DiscoveryNode node, RoutingAllocation allocation) {
-        Decision decision = shouldClusterFilter(node, allocation);
+        Decision decision = shouldIndexFilter(allocation.metadata().getIndexSafe(shardRouting.index()), node, allocation);
         if (decision != null) return decision;
 
-        decision = shouldIndexFilter(allocation.metadata().getIndexSafe(shardRouting.index()), node, allocation);
+        decision = shouldClusterFilter(node, allocation);
         if (decision != null) return decision;
 
         return allocation.decision(Decision.YES, NAME, "node passes include/exclude/require filters");
     }
 
     private Decision shouldFilter(IndexMetadata indexMd, DiscoveryNode node, RoutingAllocation allocation) {
-        Decision decision = shouldClusterFilter(node, allocation);
+        Decision decision = shouldIndexFilter(indexMd, node, allocation);
         if (decision != null) return decision;
 
-        decision = shouldIndexFilter(indexMd, node, allocation);
+        decision = shouldClusterFilter(node, allocation);
         if (decision != null) return decision;
 
         return allocation.decision(Decision.YES, NAME, "node passes include/exclude/require filters");
     }
 
     private Decision shouldIndexFilter(IndexMetadata indexMd, DiscoveryNode node, RoutingAllocation allocation) {
         if (indexMd.requireFilters() != null) {
-            if (indexMd.requireFilters().match(node) == false) {
+            if (indexMd.requireFilters().match(node)) {
+                // In case of "routing.allocation.require" prefer the index setting over the cluster setting
```

Comment:
Any reason to allow the override for `require` values only? Why not supporting it for `include` + `exclude` as well?